### ISSUE TITLE
Type checking mismatch between anonymous and named object

### ISSUE THREAD

**Issue by DorianGrey** [Microsoft/TypeScript#27938]
I'm trying to create a helper function to simplify self-assignment of values from an object provided by my applications rxjs + redux based store.
The helper function should cause a compile error if the object picked from the store contains properties that do not exist on the target object. Code and details can be found below.

**TypeScript Version:**  3.1.3
All available flags extending strictness are enabled except `strictPropertyInitialization`, and `suppressImplicitAnyIndexErrors` suppresses some errors.

**Search Terms:** type checking

**Code**
A stripped-down example of what I'm trying to achieve:
```ts
class A {
    gammel: string = "42";

    methodB() {
        // [1] Fails correctly for property "test" - does not exists on class.
        selfExtend(this)({
            gammel: "5",
            test: 14
        });

        // [2] Does not fail, even though it's the same data.
        const o = {
            gammel: "5",
            test: 14
        };
        selfExtend(this)(o);
    }
}

// Cheap trick to get around an issue with Partial<this>,
// which is acceptable in our case.
type Identity<T> = {
    [P in keyof T]: T[P];
};

export function selfExtend<T extends object>(target: Identity<T>) {
  return function(data: Partial<Identity<T>> | null) {
    if (data) {
        // In case the type-check is correct, the keys from `Object.getOwnPropertyNames`
        // should exist in both `data` and `target` - thus, assignment is safe. 
        Object.getOwnPropertyNames(data).forEach(k => (target[k] = data[k]));
    }
  };
}
```
**Expected behavior:**
In the code above, both case [1] and [2] (see comments) should cause an error that `test` does not exist on the target object.

**Actual behavior:**
Only case [1] fails, case [2] is accepted by the type checker. I.e.: Assigning the malformed value to a variable before using it causes the type check to fail, while providing the object anonymously works fine.

**Playground Link:** 
Playground can be found [here](https://www.typescriptlang.org/play/index.html#src=class%20A%20%7B%0D%0A%20%20%20%20gammel%3A%20string%20%3D%20%2242%22%3B%0D%0A%0D%0A%20%20%20%20methodB()%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20%5B1%5D%20Fails%20correctly%20for%20property%20%22test%22%20-%20does%20not%20exists%20on%20class.%0D%0A%20%20%20%20%20%20%20%20selfExtend(this)(%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20gammel%3A%20%225%22%2C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20test%3A%2014%0D%0A%20%20%20%20%20%20%20%20%7D)%3B%0D%0A%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20%5B2%5D%20Does%20not%20fail%2C%20even%20though%20it's%20the%20same%20data.%0D%0A%20%20%20%20%20%20%20%20const%20o%20%3D%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20gammel%3A%20%225%22%2C%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20test%3A%2014%0D%0A%20%20%20%20%20%20%20%20%7D%3B%0D%0A%20%20%20%20%20%20%20%20selfExtend(this)(o)%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Cheap%20trick%20to%20get%20around%20an%20issue%20with%20Partial%3Cthis%3E%2C%0D%0A%2F%2F%20which%20is%20acceptable%20in%20our%20case.%0D%0Atype%20Identity%3CT%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%3B%0D%0A%7D%3B%0D%0A%0D%0Aexport%20function%20selfExtend%3CT%20extends%20object%3E(target%3A%20Identity%3CT%3E)%20%7B%0D%0A%20%20return%20function(data%3A%20Partial%3CIdentity%3CT%3E%3E%20%7C%20null)%20%7B%0D%0A%20%20%20%20if%20(data)%20%7B%0D%0A%20%20%20%20%20%20%20%20Object.getOwnPropertyNames(data).forEach(k%20%3D%3E%20(target%5Bk%5D%20%3D%20data%5Bk%5D))%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%20).

**Related Issues:** 
Found some with similar, but not identical problems:
https://github.com/Microsoft/TypeScript/issues/27497
https://github.com/Microsoft/TypeScript/issues/26045
https://github.com/Microsoft/TypeScript/issues/17667

**Comment by jack-williams** [comment]
I think this is down to [excess property checks](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks). In [1] the literal has a contextual type from the function parameter. In [2] the literal is created without a contextual type, so there are no excess properties to look for. If you put an annotation on the declaration of `o` you get a similar error.

```ts
        const o: Partial<Identity<A>> = {
            gammel: "5",
            test: 14
//          ^^^^^^^^
//          Type '{ gammel: string; test: number; }' is not assignable to type 'Partial<Identity<A>>'.
//            Object literal may only specify known properties, and 'test' does not exist in type 'Partial<Identity<A>>'.

        };
        selfExtend(this)(o);
```

**Comment by DorianGrey** [comment]
Thanks for pointing me in this direction - didn't have the excess property check in mind.
Tried to minimize to code used for that assignment as much as possible, but it seems this need some additional checks or a different way to properly determine the types and achieve useful errors.
